### PR TITLE
(v)s(n)printf rework

### PR DESCRIPTION
**Pull request by bkentel**
- Currently the various output functions are quite wasteful w.r.t number of dynamic memory allocations, unnecessary temporaries, and copies.
- Currently, for LOCALIZE=1, builds are broken under MSVC: vsnprintf doesn't process positional parameters.
- A lot of code duplication.
- [x] Fixed MSVC via a separate code path.
- [x] Unified the general implementation for standards conformant and quirky platforms (mingw32).
- [x] Split the interface into two functions. `string_format` for code that needs and want's strings. `buffer_format` for code that only needs a buffer to pass onward.
- [x] Removed the various static, magic number sized, buffers.
- [x] Moved the base output code out of `output.*` and into `output_base.*`.
- [x] Tweaked a portion of the call-sites to use the proper implementation. TODO: many more left that could be improved by being tweaked.
- [x] Tested under MSVC (curses, sdl, localized, debug, and release) and gcc 4.8.3 via mingw-w64 (curses, sdl).
- [ ] Tested under mingw32
- [ ] Tested under linux
- [ ] The code to truncate at '\003' (needs?) to be re-added.

If anybody could test this under the missing platforms, it would be a big help.

### CODE REVIEW

**Comment by namarrgon** [comment]
It compiles without any warnings on linux with clang and gcc (with and without LOCALIZE=1).

The "Save", "Commit Suicide" and "Save Changes" dialog-windows are broken though:
http://i.imgur.com/I9ZFQBE.png?1
More of these 'confirmation' windows are affected but I assume that they are all drawn by the same code.

Apart from that I haven't found anything out of order during a short test-run.

**Comment by KA101** [comment]
mingw-w64 is not a suitable test platform.  You need mingw32 as that's what comes with the IDEs most Windows folks grab (and we the Win devs recommend.)

I'd test this, but unfortunately your last "optimization" broke things for us again.  >_<

**Comment by kevingranade** [comment]
I'm going to need to see performance impact if part of this is performance
oriented.  'Quite wasteful' isn't necessarily a reason to change anything.

**Comment by bkentel** [comment]
@namarrgon could be related to the code that did the '\003' stuff; thanks a lot for the report :smile: 

@KA101 I'll have a fix for you in a few minutes.

@kevingranade I can get some profiling metrics for you, but it's at least 2-5 fewer mallocs and memcpy's per call to the format_string. Should work out to a few ms per call. I'll post some data when I have a chance.

**Comment by bkentel** [comment]
I should say it's not **only** a performance motivated fix. The localized build wasn't functional under MSVC for one (no positional args), and it also moves the generic low level functions out into their own TU.

As for performance, before I collect some of the more detailed data, I illustrate the issue with the current set up.

For every call to, say, `mvprintz` the following must happen (for small strings):
+ (malloc) A vector of X(=1024) elements is created.
+ (memset) The entire vector is cleared.
+ (malloc) A result string is allocated.
+ (memcpy) The vector is copied to the string.
+ (free) The vector is destroyed.
+ (free) The string is destroyed.

The string itself is then never used; instead a pointer, via c_str(), to its internal buffer is passed off to the actual output routines.

With this change, the above becomes, for small (< 512) strings:
+ Grow the stack.

and for larger strings:

+ (malloc) Create a string; itself the buffer.
+ (memset) Clear the string.
+ (free) The string is destroyed.

**Comment by bkentel** [comment]
As an aside, does anyone know what the purpose of that code to truncate the string from '\003' (ETX) was for? Obviously not having it breaks the confirmation dialogs, but where else was it used, and why?

**Comment by kevingranade** [comment]
Regardless of how bad something looks, it's irrelevant without profiling
data.  If it's of the class of 'this change makes the game visibly faster'
it's one thing, but I suspect this is in the opposite extreme of being hard
to even measure.  I don't recall these coming up in profiling ever, and I'm
pretty familiar with all the game methods that take up more than 5% or so
of processing time.
That having been said I'll review it for it's cleanup component.

**Comment by bkentel** [comment]
For myself, it was a "I have to fix this anyway, so why not do the work to clean it up a bit". It may not be high-profile in the profiling data (likely partly because the code is only used when doing things that affect the UI). But all the little things **do** add up, and I think you'd agree that DDA has more than enough code for there to be lots of little things lurking in the dark corners (off the hot-path).

Regardless, thanks for taking time to give it a look. I still have to wait for some more tester feedback on the platforms I can't test, and I have to fix that prompt issue.

**Comment by bkentel** [comment]
Found the issue with the prompts -- trailing null(s) in the string interfered with concatenation.

So, nothing to do with that bit that chopped off the '\003'. No one knows what that was for? Even looking back through the history I can't find much that explains what it was.

**Comment by KA101** [comment]
Compiles w/o complaint using WinCurses, mingw32/GCC.

Unfortunately the confirmation window is borked, so it's not mergeable as it stands.

**Comment by kevingranade** [comment]
Nope, not happening.
1. vstring_format has had a lot of careful tweaking done to it, and that is being thrown away here.  I have zero confidence that a fresh implementation will be trouble-free.
2. The solution is overcomplicated by the array swapping shennanigans.  I'd need to see very strong profiling results to consider using something that convoluted.
3. The stated benefits are negligible. We're not doing localized VS builds, and this adds far more code than it removes.

I would accept a *careful* *minimal* cleanup of vstring_format, but this is not that.

**Comment by bkentel** [comment]
Did you even try the code? I can find no issues with it thus far post-confirmation dialog fix (not yet committed). There is nothing being throw away; all the original tweaking is accounted for except the '\003' which is trivial to re-add. I just wanted some idea of what it was actually for.

I could push a more minimal fix for MSVC in the meantime.

**Comment by bkentel** [comment]
Just to address your points in order:

1. I'm not asking for a merge before it has been well tested; but I have taken the original tweaking into account, and I have tested it, as have a few others. Not to say it doesn't need more testing just to be sure.
2. Efficiency comes at a cost; here it's the boilerplate needed for a discriminated union. It could be simplified a bit at the cost of one of the two use cases. UI stuff is not on the hot path, but this function is widely-used and ought to be efficient. It probably wouldn't show up in a typical profiling run because the simulation would dominate. But, if profiled while doing just UI stuff I bet it would show up in your tests.
3. It does add code. There is no denying that. The meat of it though is very small. Complexity is also another issue. This isn't something that ought to need to be extended, so I would argue maintenance costs should be minimal; further comments could help in the event that it did need to be fiddled with in the future.

**Comment by bkentel** [comment]
Finally, the implementation seems done, so I'll take another crack at reducing or at least hiding the complexity a bit.

**Comment by kevingranade** [comment]
1. Have you tested with many different languages on every platform we build for?  That's what this function has to handle, and that's the degree to which it's been tested over time.
2. That cost is unnecessary to pay if it's not needed, this is exactly what is meant by premature optimization.  If it's not on the hot path, it doesn't matter.  Making the code "more efficient" when it's not a bottleneck is meaningless, even if it doesn't make the code harder to maintain.  If it makes the code more complex it's a pure loss.  If you want to trade complexity for performance, you're going to need to show tangible runtime benefits of that performance.  Frankly you're not going to get very far with performance changes unless the issue starts with, "I noticed game slowdown in situation X, then profiled to find the problem".
3. Maintenance costs of the existing code are nonexistent, if you don't touch it there's no risk of breakage.

Bottom line, this change doesn't address anything we care about, and carries significant risk of breakage, potentially hard to spot breakage on an obscure configuration.

**Comment by bkentel** [comment]
I have a much cleaned up nearly final version I'll put up as another PR hopefully with profiling data if I can get the profiler working properly again.

What follows is less about this particular change, and more about the project's philosophy in general.

1) What are the target platforms exactly? What absolutely must build to be considered a go? Just curious here.

2) Look at it as revivification. Premature optimization is bad (especially with a looming deadline of some sort), but so too is pessimization.

"It's good enough -- don't touch it" is fine if you're working on some embedded software on a medical device that costs millions to statically verify. The cost to DDA is having someone run the code through its paces on each target, or perhaps having someone not be able to play the very latest **wip** build of master.

Honestly, pessimization and code rot (pun intended) is something I feel the code base suffers from a lot on the whole. The game is extremely sluggish in debug builds even just going through the menus. There's no reason that ought to happen; the simulation, yeah sure understandable. The UI, not so much.

There's a lot of solid, good code kicking around in the code base -- years of hard work by talented people -- but it's weighed down by a lot of code duplication and spot fixes tacked onto spot fixes rather than bolder refactors and rewrites.

Yes, something is bound to break when sweeping changes are made; perhaps some sort of limited unit testing should be slowly introduced? Some way to automate the windows builds would be a useful test for build related issues too.

**Comment by bkentel** [comment]
Ok, got the profiler working. By profiling after the game had started, just using the various menus it looks like the difference is 2.3% of samples -> 0.8% of samples, measured by summing the values of all the code calling into printstring. So, nothing astronomical, but a small improvement possibly made better by updating other code to use the correct version. The glyph lookup via std::map, and std::string::compare absolutely dominated.

In any case, I'll make a PR with the final version of the code so that it can be looked over, and work on a simpler change to get positional args working in MSVC in the mean time.

**Comment by kevingranade** [comment]
The target platforms are the release builds, currently linux curses and sdl
and windows terminal and sdl.
When something needs fixing, yes do it the right way and we just have to
deal with the fallout.  That's not what we're talking about here though,
this is a set of mostly theoretical issues that were addressed with a
rewrite of a very central piece of code.  It doesn't fix anything, so the
claim of it fighting bitrot is wrong.

You're off base about the cost of breaking things, which irritates me a
great deal.  You ignored the cost to the development team when things
break.  If a build breaks we have to jump to fix it.  If a obscure and hard
to track down bug gets introduced, someone has to hunt it down and fix it,
and that person is not necessarily you.  If that bug makes it into a
release it hurts the reputation of the project.  So no, the response of
'this doesn't fix anything, so I'm not going to touch it' is a perfectly
reasonable thing to do.

Unit testing would help somewhat with this, there's a nascent unit tests
folder which has unfortunately bitrotted, and a unfinished PR by yours
truly to get them running again.  I wasn't able to get the tests to the
point where they were worth running on every build, so of course they
bitrotted.

Re sluggish debug builds and your profiling, you might not have removed
GXXDEBUG, which swaps out all std containers for highly instrumented debug
versions of themselves, which makes operations on them up to orders of
magnitude slower.  Try turning that off and seeing how sluggish things are,
and turning that off is a must for proper profiling.

**Comment by DavidKeaton** [comment]
Is it simply disabling this env var?

**Comment by kevingranade** [comment]
It's being set in the makefile, one of the earliest non-comment lines.

**Comment by bkentel** [comment]
Fair enough about breaking things. I realize the potential cost to other devs and don't mean to downplay that at all, just that there is a balance to strike between pessimization and optimization. Here perhaps it's not warranted; I made a PR for a much simpler change. Not theoretically as performant, but much easier to reason about -- fewer moving parts.

I'm glad to hear that unit testing in some form is at least on the table and that it was you championing it. Hopefully it can be resurrected or reincarnated as something else at some point.

Re: sluggish builds. For debugging I absolutely want the paranoid version of the containers, but it's much slower than I would expect based on past projects.

Re: profiling. The results I collected were for full-optimization with link-time code generation also turned on; basically as good as it can get without doing profile-guided optimization. The results for these changes however are hard to confidently assert are more than line noise level though.It's absolutely drowned by an order of magnitude by rendering code deeper down.

Thanks, by the way, for taking the time to explain your position.

**Comment by DavidKeaton** [comment]
Ah okay, thanks Kevin!
